{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dbf45a8-680d-4931-abb4-745f52d078d3",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Base imports\n",
    "import os\n",
    "import pickle\n",
    "\n",
    "# Compute imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from tqdm.notebook import tqdm, trange\n",
    "\n",
    "# Plotting imports\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "import seaborn as sns\n",
    "from plotly import express as px\n",
    "\n",
    "# ML import\n",
    "from sklearn.decomposition import NMF\n",
    "from sklearn.metrics import mean_squared_error, median_absolute_error\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "\n",
    "import multiprocessing\n",
    "from multiprocessing import Pool\n",
    "\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['svg.fonttype'] = 'none'\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Arial'\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "sns.set_style('ticks')\n",
    "matplotlib.rcParams['text.color'] = '#000000'\n",
    "matplotlib.rcParams['axes.labelcolor'] = '#000000'\n",
    "matplotlib.rcParams['xtick.color'] = '#000000'\n",
    "matplotlib.rcParams['ytick.color'] = '#000000'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c541d8-9901-4d52-81fc-7d60fac22af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GENES = '../../data/processed/cd-hit-results/sim80/Ebacter_strain_by_gene.pickle.gz'\n",
    "ENRICHED_METADATA = '../../data/metadata/enriched_metadata.csv'\n",
    "DF_EGGNOG = '../../data/processed/df_eggnog.csv'\n",
    "\n",
    "DF_ACC_COMPLETE = '../../data/processed/CAR_genomes/df_acc_complete.pickle'\n",
    "DF_RARE_COMPLETE = '../../data/processed/CAR_genomes/df_rare_complete.pickle'\n",
    "DF_CORE_COMPLETE = '../../data/processed/CAR_genomes/df_core_complete.pickle'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "410d99fa-70b9-429d-b4b0-793424cb09e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rare = pd.read_pickle(DF_RARE_COMPLETE)\n",
    "df_acc = pd.read_pickle(DF_ACC_COMPLETE)\n",
    "df_core = pd.read_pickle(DF_CORE_COMPLETE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e57a713-1a7b-4967-ab51-6acc0ba0dd0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "A_BINARIZED = '../../data/processed/nmf-outputs/A_binarized.csv'\n",
    "A_binarized = pd.read_csv(A_BINARIZED, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87fb9379-8730-4209-b183-b2c17a6d269b",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(ENRICHED_METADATA, index_col=0, dtype='object')\n",
    "\n",
    "display( metadata.shape, metadata.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6326fa4-7ad5-4f4a-bd8d-827a10455250",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in (full) P matrix\n",
    "df_genes = pd.read_pickle(DF_GENES)\n",
    "\n",
    "# Filter metadata for Complete sequences only\n",
    "metadata_complete = metadata[metadata.genome_status == 'Complete'] # filter for only Complete sequences\n",
    "\n",
    "# Filter P matrix for Complete sequences only\n",
    "df_genes_complete = df_genes[metadata_complete.genome_id].copy()\n",
    "df_genes_complete.fillna(0, inplace=True) # replace N/A with 0\n",
    "df_genes_complete = df_genes_complete.sparse.to_dense().astype('int8') # densify & typecast to int8 for space and compute reasons\n",
    "inCompleteseqs = df_genes_complete.sum(axis=1) > 0 # filter for genes found in complete sequences\n",
    "df_genes_complete = df_genes_complete[inCompleteseqs]\n",
    "\n",
    "df_genes_complete.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30992e3b-8c65-4543-84c6-6deb5cf21bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in eggNOG annotations\n",
    "df_eggnog = pd.read_csv(DF_EGGNOG, index_col=0)\n",
    "df_eggnog.fillna('-', inplace=True)\n",
    "\n",
    "display(\n",
    "    df_eggnog.shape,\n",
    "    df_eggnog.head()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44ead6e4-d434-4302-993d-dc41d21ae6f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "P_allele = pd.read_pickle('../../data/processed/cd-hit-results/sim80/Ebacter_strain_by_allele.pickle.gz')\n",
    "P_allele = P_allele.fillna(0).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3cfbf19-76be-44c7-aa7e-626813bd4ff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes = df_core.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cee4947-ebdf-4d75-aaea-f076e038baca",
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_genes = df_acc.index\n",
    "rare_genes = df_rare.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c40ce94-fb78-4189-8a76-98852b194844",
   "metadata": {},
   "outputs": [],
   "source": [
    "characterized_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']\n",
    "\n",
    "\n",
    "df = A_binarized.loc[characterized_order]\n",
    "\n",
    "# List to store labels and column names\n",
    "label_col = []\n",
    "name_col = []\n",
    "\n",
    "# Iterate over columns\n",
    "for col in df.columns:\n",
    "    # Find index where value is 1\n",
    "    index = df.index[df[col] == 1].tolist()\n",
    "    if index:\n",
    "        label_col.append(index[0])  # Append the first index where value is 1\n",
    "    else:\n",
    "        label_col.append('None')  # If no 1 is found, append None\n",
    "    name_col.append(col)\n",
    "\n",
    "# Create a new DataFrame\n",
    "output_df = pd.DataFrame({'Column': name_col, 'Label': label_col}).set_index('Column')\n",
    "\n",
    "custom_colors = [\n",
    "    # Shades of red/orange/yellow\n",
    "    \"Red\",\n",
    "    \"IndianRed\",\n",
    "    \"DarkRed\",\n",
    "    \"FireBrick\",\n",
    "    \"Tomato\",\n",
    "    \"Gold\",\n",
    "    \"DarkGoldenrod\",\n",
    "    \"Goldenrod\",\n",
    "    # Other species\n",
    "    \"Green\",\n",
    "    \"Blue\",\n",
    "    \"Purple\",\n",
    "    \"Cyan\",\n",
    "    \"Magenta\",\n",
    "    \"Lime\",\n",
    "    \"Pink\",\n",
    "]\n",
    "\n",
    "\n",
    "clr = dict(zip(characterized_order + [\"None\"], custom_colors))\n",
    "output_df['color'] = output_df.Label.map(clr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f71577e0-9e99-4ece-8fde-15c399c2f21e",
   "metadata": {},
   "source": [
    "## Extract Allele Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d72870ff-c661-4bc5-bccc-9adfe4e55b8f",
   "metadata": {},
   "source": [
    "The below is based on an old method I used but needs to be changed or re-worked. I need to do the following:\n",
    "1. Find software to build a phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86f8d502-367f-4f69-9e98-d7fabfb2958c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "from io import StringIO\n",
    "from Bio import SeqIO\n",
    "import tempfile\n",
    "from Bio import AlignIO\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2570e0dd-4fc7-4a02-b1be-95bee679e09e",
   "metadata": {},
   "source": [
    "# Analysis for Single Copy Core Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e53ca4-1c13-4472-84cc-2899d7f46b49",
   "metadata": {},
   "outputs": [],
   "source": [
    "scc_genes = ['Ebacter_C100472',\n",
    " 'Ebacter_C101558',\n",
    " 'Ebacter_C103292',\n",
    " 'Ebacter_C10426',\n",
    " 'Ebacter_C107942',\n",
    " 'Ebacter_C108573',\n",
    " 'Ebacter_C110315',\n",
    " 'Ebacter_C112797',\n",
    " 'Ebacter_C113365',\n",
    " 'Ebacter_C114601',\n",
    " 'Ebacter_C11866',\n",
    " 'Ebacter_C12400',\n",
    " 'Ebacter_C13432',\n",
    " 'Ebacter_C13437',\n",
    " 'Ebacter_C13742',\n",
    " 'Ebacter_C13804',\n",
    " 'Ebacter_C13806',\n",
    " 'Ebacter_C14408',\n",
    " 'Ebacter_C16067',\n",
    " 'Ebacter_C16181',\n",
    " 'Ebacter_C17612',\n",
    " 'Ebacter_C18127',\n",
    " 'Ebacter_C19902',\n",
    " 'Ebacter_C21434',\n",
    " 'Ebacter_C2440',\n",
    " 'Ebacter_C24576',\n",
    " 'Ebacter_C25412',\n",
    " 'Ebacter_C25641',\n",
    " 'Ebacter_C26119',\n",
    " 'Ebacter_C27162',\n",
    " 'Ebacter_C27467',\n",
    " 'Ebacter_C28628',\n",
    " 'Ebacter_C28733',\n",
    " 'Ebacter_C29374',\n",
    " 'Ebacter_C30093',\n",
    " 'Ebacter_C34271',\n",
    " 'Ebacter_C34511',\n",
    " 'Ebacter_C3501',\n",
    " 'Ebacter_C35293',\n",
    " 'Ebacter_C35487',\n",
    " 'Ebacter_C37335',\n",
    " 'Ebacter_C37771',\n",
    " 'Ebacter_C38778',\n",
    " 'Ebacter_C38917',\n",
    " 'Ebacter_C39435',\n",
    " 'Ebacter_C42042',\n",
    " 'Ebacter_C4210',\n",
    " 'Ebacter_C42587',\n",
    " 'Ebacter_C43046',\n",
    " 'Ebacter_C43535',\n",
    " 'Ebacter_C43888',\n",
    " 'Ebacter_C44386',\n",
    " 'Ebacter_C44907',\n",
    " 'Ebacter_C48305',\n",
    " 'Ebacter_C49588',\n",
    " 'Ebacter_C49697',\n",
    " 'Ebacter_C49722',\n",
    " 'Ebacter_C50924',\n",
    " 'Ebacter_C51224',\n",
    " 'Ebacter_C51497',\n",
    " 'Ebacter_C51498',\n",
    " 'Ebacter_C51904',\n",
    " 'Ebacter_C52224',\n",
    " 'Ebacter_C58596',\n",
    " 'Ebacter_C59704',\n",
    " 'Ebacter_C59913',\n",
    " 'Ebacter_C61078',\n",
    " 'Ebacter_C61459',\n",
    " 'Ebacter_C61836',\n",
    " 'Ebacter_C6228',\n",
    " 'Ebacter_C62596',\n",
    " 'Ebacter_C62952',\n",
    " 'Ebacter_C63161',\n",
    " 'Ebacter_C65930',\n",
    " 'Ebacter_C67022',\n",
    " 'Ebacter_C67092',\n",
    " 'Ebacter_C67503',\n",
    " 'Ebacter_C69060',\n",
    " 'Ebacter_C70584',\n",
    " 'Ebacter_C7096',\n",
    " 'Ebacter_C71129',\n",
    " 'Ebacter_C71264',\n",
    " 'Ebacter_C72288',\n",
    " 'Ebacter_C73125',\n",
    " 'Ebacter_C73519',\n",
    " 'Ebacter_C74659',\n",
    " 'Ebacter_C74690',\n",
    " 'Ebacter_C75003',\n",
    " 'Ebacter_C76258',\n",
    " 'Ebacter_C78689',\n",
    " 'Ebacter_C80576',\n",
    " 'Ebacter_C81013',\n",
    " 'Ebacter_C81014',\n",
    " 'Ebacter_C82241',\n",
    " 'Ebacter_C85970',\n",
    " 'Ebacter_C88246',\n",
    " 'Ebacter_C91317',\n",
    " 'Ebacter_C91821',\n",
    " 'Ebacter_C92829',\n",
    " 'Ebacter_C9312',\n",
    " 'Ebacter_C93270',\n",
    " 'Ebacter_C94777',\n",
    " 'Ebacter_C95738',\n",
    " 'Ebacter_C96274',\n",
    " 'Ebacter_C97800']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83280c16-dacd-4129-8b69-13ae405aa246",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(scc_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce530953-b9cc-4608-93cb-bdcfcd206346",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set gene of interest for multiple alignment\n",
    "gene = scc_genes\n",
    "alleles = [x for x in P_allele.index if x.split('A')[0] in gene]\n",
    "allele_P = P_allele.loc[alleles, df_core.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f2e6e6-0d15-441b-9b33-0131ae667bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the FASTA file\n",
    "fasta_file = \"../../data/processed/cd-hit-results/sim80/Ebacter_nr.faa\"\n",
    "\n",
    "\n",
    "# Dictionary to store sequences\n",
    "sequences = {}\n",
    "\n",
    "# Iterate through the FASTA file and store sequences associated with headers in gene_list\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    if record.id in allele_P.index:\n",
    "        sequences[record.id] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03dc071b-bef8-488f-a67d-5b25989aa042",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_seqeunces = {}\n",
    "for strain in tqdm(allele_P.columns):\n",
    "    strain_seqeunces[strain] = \"\"\n",
    "    for target_seq in allele_P[strain][allele_P[strain] > 0].index:\n",
    "        strain_seqeunces[strain] += sequences[target_seq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0dfc60f-7da2-436c-8bd1-7b3f83116817",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"strains_sequences.fasta\", \"w\") as fasta_file:\n",
    "    for strain, sequence in strain_seqeunces.items():\n",
    "        # Write the strain name as the header line and the sequence as the sequence line\n",
    "        fasta_file.write(f\"> {strain}\\n{sequence}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550c31d3-2b56-4235-9d73-f343f10fd702",
   "metadata": {},
   "source": [
    "I ran command line mafft on the fasta file using the following command: \n",
    "```\n",
    " # old mafft --thread 60 --algorithm FFT-NS-2 strains_sequences.fasta > aligned_sequences.fasta\n",
    "\"/home/jtburrows/private/phylon_analysis/bin/mafft\"  --retree 2 --inputorder \"strains_sequences.fasta\" > \"aligned_sequences.fasta\"\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f352c98-3ed9-426c-89ba-514f232aeb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the alignment\n",
    "alignment = AlignIO.read(('aligned_sequences.fasta'), \"fasta\")\n",
    "\n",
    "# # Save the alignment in Clustal format (optional)\n",
    "# with open(\"alignment.aln\", \"w\") as output_file:\n",
    "#     AlignIO.write(alignment, output_file, \"clustal\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e92505d-2c3c-405a-b1bc-16e24f8899ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the distance matrix\n",
    "calculator = DistanceCalculator('identity')\n",
    "distance_matrix = calculator.get_distance(alignment)\n",
    "\n",
    "# Build the tree using the neighbor-joining algorithm\n",
    "constructor = DistanceTreeConstructor()\n",
    "tree = constructor.nj(distance_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16d6bdcb-5ad4-4e95-b49e-26be22d7c50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from io import StringIO\n",
    "\n",
    "# Assuming you have your tree object in the variable `tree`\n",
    "output = StringIO()\n",
    "Phylo.write(tree, output, \"newick\")\n",
    "newick_str = output.getvalue()\n",
    "\n",
    "# Save to a file\n",
    "with open(\"tree.newick\", \"w\") as f:\n",
    "    f.write(newick_str)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "747aefee-f4ab-48a8-a658-cbaf93724996",
   "metadata": {},
   "source": [
    "# Plot as a circular tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e64dae6-b352-4823-8f25-9a52da819ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_species = metadata_complete[metadata_complete.genome_status == 'Complete'].loc[:,[\"genome_id\", \"genome_name\"]]\n",
    "df_species[\"species\"] = df_species[\"genome_name\"].apply(lambda x: x.split()[0]+\" \" +x.split()[1])\n",
    "df_species.set_index('genome_id', inplace=True)\n",
    "custom_colors = {'Enterobacter hormaechei': 'FireBrick',\n",
    " 'Enterobacter cloacae': 'Pink',\n",
    " 'Enterobacter sp.': 'SlateGray',\n",
    " 'Enterobacter roggenkampii': 'Green',\n",
    " 'Enterobacter kobei': 'Purple',\n",
    " 'Enterobacter cancerogenus': 'Magenta',\n",
    " 'Enterobacter bugandensis': 'Cyan',\n",
    " 'Enterobacter asburiae': 'Blue',\n",
    " 'Enterobacter ludwigii': 'Lime',\n",
    " 'Enterobacter mori': '#F5F5DC',\n",
    " 'Enterobacter xiangfangensis': 'Red'}\n",
    "\n",
    "df_species['color'] = df_species.species.map(custom_colors)\n",
    "\n",
    "if 'ref_seq' in strain_seqeunces.keys():\n",
    "    df_species.loc['ref'] = ['ref', 'ref', 'black']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6cbb301-ff1f-49f9-816a-66e6c030fec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_strains(phylon, A_binarized = A_binarized):\n",
    "    phylon_membership = A_binarized.loc[phylon]\n",
    "    return (phylon_membership[phylon_membership == 1]).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f2266a-17c3-4f7c-9622-d40f6946dfd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alternate colors for each strain from phylons\n",
    "characterized_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']\n",
    "\n",
    "df = A_binarized.loc[characterized_order]\n",
    "\n",
    "# List to store labels and column names\n",
    "label_col = []\n",
    "name_col = []\n",
    "\n",
    "# Iterate over columns\n",
    "for col in df.columns:\n",
    "    # Find index where value is 1\n",
    "    index = df.index[df[col] == 1].tolist()\n",
    "    if index:\n",
    "        label_col.append(index[0])  # Append the first index where value is 1\n",
    "    else:\n",
    "        label_col.append('None')  # If no 1 is found, append None\n",
    "    name_col.append(col)\n",
    "\n",
    "# Create a new DataFrame\n",
    "output_df = pd.DataFrame({'Column': name_col, 'Label': label_col}).set_index('Column')\n",
    "\n",
    "custom_colors = [\n",
    "    # Shades of red/orange/yellow\n",
    "    \"#FF0000\",      # Red\n",
    "    \"#CD5C5C\",      # IndianRed\n",
    "    \"#8B0000\",      # DarkRed\n",
    "    \"#B22222\",      # FireBrick\n",
    "    \"#FF6347\",      # Tomato\n",
    "    \"#FFD700\",      # Gold\n",
    "    \"#B8860B\",      # DarkGoldenrod\n",
    "    \"#DAA520\",      # Goldenrod\n",
    "    # Other species\n",
    "    \"#008000\",      # Green\n",
    "    \"#0000FF\",      # Blue\n",
    "    \"#800080\",      # Purple\n",
    "    \"#00FFFF\",      # Cyan\n",
    "    \"#FF00FF\",      # Magenta\n",
    "    \"#00FF00\",      # Lime\n",
    "    \"#FFC0CB\",      # Pink\n",
    "]\n",
    "\n",
    "clr = dict(zip(characterized_order + [\"None\"], custom_colors))\n",
    "custom_colors = clr\n",
    "output_df['color'] = output_df.Label.map(clr)\n",
    "\n",
    "df_species = output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d0ea44-62de-41d4-bc3b-ca827a653dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "L_BINARIZED = '../../data/processed/nmf-outputs/L_binarized.csv'\n",
    "A_BINARIZED = '../../data/processed/nmf-outputs/A_binarized.csv'\n",
    "\n",
    "# Load in L_binarized matrix\n",
    "L_binarized = pd.read_csv(L_BINARIZED, index_col=0)\n",
    "A_binarized = pd.read_csv(A_BINARIZED, index_col=0)\n",
    "L_binarized[(L_binarized.loc[:,'hormaechei-steigerwaltii-3'] == 1) & (L_binarized.loc[:,'hormaechei-steigerwaltii-1'] == 1) & (L_binarized.sum(axis=1) == 2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "992ee958-b8e3-45bb-809e-ec0f0160d60b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from pycirclize import Circos\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "# Initialize the Circos plot\n",
    "circos, tv = Circos.initialize_from_tree(\"tree.newick\", \n",
    "    leaf_label_size=0,\n",
    "    start=10,\n",
    "    end=350,\n",
    "    r_lim=(20, 55),\n",
    "    line_kws=dict(color=\"black\", lw=2),\n",
    "    align_line_kws=dict(ls=\"dashdot\", lw=.3, alpha=.5),\n",
    "    align_leaf_label=True,\n",
    "    ladderize=True\n",
    ")\n",
    "\n",
    "\n",
    "# Plot heatmap with various style\n",
    "sector = circos.sectors[0]\n",
    "sector.rect(r_lim=(55, 65), ec=\"grey\", lw=1)\n",
    "\n",
    "# get labels for the phylons to map to the phylon colors\n",
    "label_encoder = LabelEncoder()\n",
    "test_df = df_species.copy()\n",
    "test_df['num_labels'] = label_encoder.fit_transform(df_species.Label)\n",
    "\n",
    "heatmap_track1 = sector.add_track((55, 60))\n",
    "phylon_colors = LinearSegmentedColormap.from_list(\"phylon_colors\", test_df.loc[tv.leaf_labels].sort_values('num_labels').fillna('white').color.unique())\n",
    "heatmap_track1.heatmap(test_df.loc[tv.leaf_labels].num_labels.values, cmap=phylon_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"Phylon\", r=heatmap_track1.r_center, size=8, color=\"black\")\n",
    "\n",
    "# list of colors for the mash clusters\n",
    "color_list = [\n",
    "    \"#FFFF00\",  # Bright Yellow\n",
    "    \"#FFEA00\",  # Yellow 1 (brighter yellow)\n",
    "    \"#FFD700\",  # Gold (distinct yellow)\n",
    "    \"#FF5500\",  # Deep Yellow-Orange\n",
    "    \"#FF9900\",  # Deep Orange 1\n",
    "    \"#FF8000\",  # Orange 1\n",
    "    \"#FF6600\",  # Orange 2\n",
    "    \"#FF3300\",  # Deep Orange-Red\n",
    "    \"#FF4444\",  # Red\n",
    "    \"#9F7FBF\",  # Purple\n",
    "    \"#66A066\",  # Green 1\n",
    "    \"#66D966\",  # Green 2\n",
    "    \"#6699FF\",  # Blue 1\n",
    "    \"#80BFFF\",  # Blue 2\n",
    "    \"#80A3C1\",  # Blue 3\n",
    "    \"#A3B8C0\",  # Blue 4\n",
    "    \"#F2F2F2\",  # White\n",
    "    \"#80B3B3\",  # Teal\n",
    "    \"#B3FFB3\",  # Lime\n",
    "    \"#C2A8E0\",  # Bright Purple\n",
    "    \"#FF80BF\"   # Pink\n",
    "]\n",
    "\n",
    "data = metadata_complete.set_index('genome_id')\n",
    "heatmap_track2 = sector.add_track((60, 65))\n",
    "mash_cluster_colors = LinearSegmentedColormap.from_list(\"mash_cluster_colors\", color_list)\n",
    "heatmap_track2.heatmap(data.loc[tv.leaf_labels].complete_mash_cluster.values.astype(float), cmap=mash_cluster_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"Mash Cluster\", r=heatmap_track2.r_center, size=8, color=\"black\")\n",
    "\n",
    "### add row for traits of interest ###\n",
    "\n",
    "## mobile phylon tracks\n",
    "phylon_data = (A_binarized.loc['unchar-1', tv.leaf_labels] > 0).astype(int)\n",
    "heatmap_mobile1 = sector.add_track((66, 68))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#FF6961\"])\n",
    "heatmap_mobile1.heatmap(phylon_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"mobile-1\", r=heatmap_mobile1.r_center, size=8, color=\"black\")\n",
    "\n",
    "phylon_data = (A_binarized.loc['unchar-2', tv.leaf_labels] > 0).astype(int)\n",
    "heatmap_mobile2 = sector.add_track((68, 70))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#AEC6CF\"])\n",
    "heatmap_mobile2.heatmap(phylon_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"mobile-2\", r=heatmap_mobile2.r_center, size=8, color=\"black\")\n",
    "\n",
    "phylon_data = (A_binarized.loc['unchar-3', tv.leaf_labels] > 0).astype(int)\n",
    "heatmap_mobile3 = sector.add_track((70, 72))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#77DD77\"])\n",
    "heatmap_mobile3.heatmap(phylon_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"mobile-3\", r=heatmap_mobile3.r_center, size=8, color=\"black\")\n",
    "\n",
    "## mobile phylon tracks\n",
    "phylon_data = (A_binarized.loc['unchar-4', tv.leaf_labels] > 0).astype(int)\n",
    "heatmap_mobile4 = sector.add_track((72, 74))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#B39EB5\"])\n",
    "heatmap_mobile4.heatmap(phylon_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"mobile-4\", r=heatmap_mobile4.r_center, size=8, color=\"black\")\n",
    "\n",
    "\n",
    "# trait 1\n",
    "# pqq genes associated with pyrroloquinoline quinone\n",
    "# only found in hormaechei cluster\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2245851/ - promotes plant growth, rhizosphere associated\n",
    "cond = df_eggnog.loc[df_genes.index].Preferred_name.apply(lambda x : 'pqq' in x)\n",
    "cond2 = df_eggnog.loc[df_genes.index][cond].index\n",
    "inds = df_acc.loc[df_eggnog.loc[[x for x in cond2 if x in df_acc.index]].index, tv.leaf_labels].index\n",
    "genetic_trait_data = (df_acc.loc[inds, tv.leaf_labels].sum() > 2).astype(int)\n",
    "heatmap_track3 = sector.add_track((75, 76))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#FF0000\"])\n",
    "heatmap_track3.heatmap(genetic_trait_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"pqq Genes\", r=heatmap_track3.r_center, size=8, color=\"black\")\n",
    "\n",
    "# trait 2\n",
    "# genes associated with Salmochelin production\n",
    "cond = df_eggnog.loc[df_genes.index].Preferred_name.apply(lambda x : 'iro' in x)\n",
    "cond2 = df_eggnog.loc[df_genes.index][cond].index\n",
    "inds = df_genes_complete.loc[df_eggnog.loc[[x for x in cond2 if x in df_genes_complete.index]].index, tv.leaf_labels].index\n",
    "genetic_trait_data = (df_genes_complete.loc[inds, tv.leaf_labels].sum() > 0).astype(int)\n",
    "heatmap_track4 = sector.add_track((77, 78))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#000000\"])\n",
    "heatmap_track4.heatmap(genetic_trait_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"Salmochelin VFs\", r=heatmap_track4.r_center, size=8, color=\"black\")\n",
    "\n",
    "# trait 3, arn operon associated with colistin resistance\n",
    "cond = df_eggnog.loc[df_genes.index].Preferred_name.apply(lambda x : 'arn' in x)\n",
    "cond2 = df_eggnog.loc[df_genes.index][cond].index\n",
    "inds = df_acc.loc[df_eggnog.loc[[x for x in cond2 if x in df_acc.index]].index, tv.leaf_labels].index\n",
    "genetic_trait_data = (df_acc.loc[inds, tv.leaf_labels].sum() > 6).astype(int)\n",
    "heatmap_track5 = sector.add_track((79, 80))\n",
    "trait_colors = LinearSegmentedColormap.from_list(\"trait_colors\", [\"#FFFFFF\", \"#3333FF\"])\n",
    "heatmap_track5.heatmap(genetic_trait_data.loc[tv.leaf_labels].values, cmap=trait_colors, rect_kws=dict(ec=\"lightgrey\", lw=0))\n",
    "circos.text(\"Arn Operon\", r=heatmap_track5.r_center, size=8, color=\"black\")\n",
    "\n",
    "\n",
    "### End of adding traits of interest ###\n",
    "# Create the Circos figure\n",
    "fig = circos.plotfig(figsize=(20,10))\n",
    "\n",
    "# Plot legend for phylons\n",
    "line_handles = []\n",
    "\n",
    "for label, name, color in zip(test_df.loc[tv.leaf_labels].num_labels.unique(),test_df.loc[tv.leaf_labels].Label.unique(), test_df.loc[tv.leaf_labels].color.fillna('white').unique()):\n",
    "    line_handles.append(Line2D([], [], color=color, label=name, lw=4))\n",
    "\n",
    "# line_legend = circos.ax.legend(\n",
    "#     handles=line_handles,\n",
    "#     bbox_to_anchor=(0.80, 1.02),\n",
    "#     # loc=\"upper right\",\n",
    "#     fontsize=8,\n",
    "#     title=\"Phylons\",\n",
    "#     handlelength=2,\n",
    "#     ncols=1\n",
    "# )\n",
    "# circos.ax.add_artist(line_legend)\n",
    "\n",
    "# # Plot legend for mash cluster\n",
    "# line_handles = []\n",
    "\n",
    "# for name, color in zip(list(range(int(data.loc[tv.leaf_labels].complete_mash_cluster.astype(float).max()))), color_list):\n",
    "#     line_handles.append(Line2D([], [], color=color, label=name, lw=4))\n",
    "\n",
    "# line_legend = circos.ax.legend(\n",
    "#     handles=line_handles,\n",
    "#     bbox_to_anchor=(0, 1.02),\n",
    "#     loc=\"upper left\",\n",
    "#     fontsize=8,\n",
    "#     title=\"Mash Clusters\",\n",
    "#     handlelength=2,\n",
    "#     ncols=1\n",
    "# )\n",
    "# circos.ax.add_artist(line_legend)\n",
    "\n",
    "# plt.title(\"Single-Copy Core Gene Phylogeny\")\n",
    "plt.savefig(\"circos_plot.png\", format='png', dpi=600, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e386ee-11db-4037-bf14-18ed3cc46f61",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:phylon_analysis]",
   "language": "python",
   "name": "conda-env-phylon_analysis-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
