{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4d3b588-27f0-41f6-89b3-186d33aefdf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Base imports\n",
    "import os\n",
    "import pickle\n",
    "import re\n",
    "\n",
    "# Compute imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from tqdm.notebook import tqdm, trange\n",
    "\n",
    "# Plotting imports\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotly import express as px\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "# ML import\n",
    "from sklearn.decomposition import NMF\n",
    "from sklearn.metrics import mean_squared_error, median_absolute_error\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['svg.fonttype'] = 'none'\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Arial'\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "sns.set_style('ticks')\n",
    "matplotlib.rcParams['text.color'] = '#000000'\n",
    "matplotlib.rcParams['axes.labelcolor'] = '#000000'\n",
    "matplotlib.rcParams['xtick.color'] = '#000000'\n",
    "matplotlib.rcParams['ytick.color'] = '#000000'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee345c05-568a-4aa8-9798-c9d9a93cd8b1",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "DF_GENES = '../../data/processed/cd-hit-results/sim80/Ebacter_strain_by_gene.pickle.gz'\n",
    "ENRICHED_METADATA = '../../data/metadata/enriched_metadata.csv'\n",
    "DF_EGGNOG = '../../data/processed/df_eggnog.csv'\n",
    "\n",
    "DF_CORE_COMPLETE = '../../data/processed/CAR_genomes/df_core_complete.pickle'\n",
    "DF_ACC_COMPLETE = '../../data/processed/CAR_genomes/df_acc_complete.pickle'\n",
    "DF_RARE_COMPLETE = '../../data/processed/CAR_genomes/df_rare_complete.pickle'\n",
    "\n",
    "L_BINARIZED = '../../data/processed/nmf-outputs/L_binarized.csv'\n",
    "A_BINARIZED = '../../data/processed/nmf-outputs/A_binarized.csv'\n",
    "L_MATRIX = '../../data/processed/nmf-outputs/L.csv'\n",
    "A_MATRIX = '../../data/processed/nmf-outputs/A.csv'\n",
    "BAKTA_ANNOTATIONS = '../../data/processed/bakta_gene_annotations.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8d08ff1-68b7-4c32-8476-a196b8a878a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_annotations = pd.read_csv(BAKTA_ANNOTATIONS, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96e31e5-d72f-4987-bbd5-bfe9e78bd667",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_locs_acc = pd.read_csv('../acc_gene_location.csv', index_col=0)\n",
    "gene_locs = pd.read_csv('../complete_gene_location.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "243ed8ca-90f8-4f32-aaec-25fd21551445",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rare = pd.read_pickle(DF_RARE_COMPLETE)\n",
    "df_acc = pd.read_pickle(DF_ACC_COMPLETE)\n",
    "df_core = pd.read_pickle(DF_CORE_COMPLETE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9da9a1a-720f-4298-ae13-283be2375ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(ENRICHED_METADATA, index_col=0, dtype='object')\n",
    "\n",
    "display( metadata.shape, metadata.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7e566a-8c82-475b-8bab-61263159d91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in (full) P matrix\n",
    "df_genes = pd.read_pickle(DF_GENES)\n",
    "\n",
    "# Filter metadata for Complete sequences only\n",
    "metadata_complete = metadata[metadata.genome_status == 'Complete'] # filter for only Complete sequences\n",
    "\n",
    "# Filter P matrix for Complete sequences only\n",
    "df_genes_complete = df_genes[metadata_complete.genome_id].copy()\n",
    "df_genes_complete.fillna(0, inplace=True) # replace N/A with 0\n",
    "df_genes_complete = df_genes_complete.sparse.to_dense().astype('int8') # densify & typecast to int8 for space and compute reasons\n",
    "inCompleteseqs = df_genes_complete.sum(axis=1) > 0 # filter for genes found in complete sequences\n",
    "df_genes_complete = df_genes_complete[inCompleteseqs]\n",
    "\n",
    "df_genes_complete.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3248cc6e-8de0-4eaa-8d03-d9cd2f6efa32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in eggNOG annotations\n",
    "df_eggnog = pd.read_csv(DF_EGGNOG, index_col=0)\n",
    "df_eggnog.fillna('-', inplace=True)\n",
    "\n",
    "display(\n",
    "    df_eggnog.shape,\n",
    "    df_eggnog.head()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e84e67f3-e5fc-4f35-8fb0-4be037abbc3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in A_binarized matrix\n",
    "A_binarized = pd.read_csv(A_BINARIZED, index_col=0)\n",
    "A_binarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb7e8a8-b8a2-4eee-9659-3cc212a68a01",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load in L_binarized matrix\n",
    "L_binarized = pd.read_csv(L_BINARIZED, index_col=0)\n",
    "L_binarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab25cb48-aac6-4fd2-af6b-cd7403b3c91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "phylon_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'unchar-1',\n",
    " 'unchar-2',\n",
    " 'unchar-3',\n",
    " 'unchar-4',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']\n",
    "\n",
    "characterized_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87814f67-759a-4939-946c-00aa47a63960",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene_order = []\n",
    "\n",
    "# Add in zero-phylon genes\n",
    "zero_cond = L_binarized.sum(axis=1) == 0\n",
    "gene_order.extend(L_binarized[zero_cond].index)\n",
    "\n",
    "# Add in single-phylon genes\n",
    "for phylon in phylon_order:\n",
    "    single_cond = L_binarized.sum(axis=1) == 1\n",
    "    inPhylon = L_binarized[phylon] == 1\n",
    "    gene_order.extend(L_binarized[inPhylon & single_cond].index)\n",
    "\n",
    "# Add in poly-phylon genes\n",
    "for num_active_phylons in trange(2, int(L_binarized.sum(axis=1).max())+1):\n",
    "    num_cond = L_binarized.sum(axis=1) == num_active_phylons\n",
    "    gg = sns.clustermap(L_binarized[num_cond], method='ward', metric='euclidean', col_cluster=False, yticklabels=False);\n",
    "    gene_order.extend(gg.data2d.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84e2690-209c-48cb-be58-2450ad4187ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main sorted clustermap\n",
    "\n",
    "g = sns.clustermap(\n",
    "    L_binarized.loc[gene_order],\n",
    "    method='ward',\n",
    "    metric='euclidean',\n",
    "    row_cluster=False,\n",
    "    yticklabels=False,\n",
    "    cmap='Greys'\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24abfe04-e4db-46da-805d-40df26238f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_order = []\n",
    "unchar_strain_order = []\n",
    "\n",
    "\n",
    "# zero-phylon strains\n",
    "noPhylon = A_binarized.sum() == 0\n",
    "strain_order.extend(A_binarized.sum()[noPhylon].index.tolist())\n",
    "\n",
    "# strain lists\n",
    "single_phylon_strains = A_binarized.sum()[A_binarized.sum() == 1].index\n",
    "multi_phylon_strains = A_binarized.sum()[A_binarized.sum() > 1].index\n",
    "\n",
    "for phylon in phylon_order:\n",
    "    if 'unchar' in phylon:\n",
    "        continue\n",
    "    else:\n",
    "        phylon_aff_binarized_single = A_binarized.loc[phylon, single_phylon_strains]\n",
    "        phylon_aff_binarized_multi = A_binarized.loc[phylon, multi_phylon_strains]\n",
    "    \n",
    "        inPhylon_single = phylon_aff_binarized_single == 1\n",
    "        inPhylon_multi = phylon_aff_binarized_multi == 1\n",
    "    \n",
    "        list1 = phylon_aff_binarized_single[inPhylon_single].index.tolist()\n",
    "        list2 = phylon_aff_binarized_multi[inPhylon_multi].index.tolist()\n",
    "        new_list2 = list(set(list2) - set(strain_order)) # ensures no double-counting\n",
    "        \n",
    "        strain_order.extend(list1)\n",
    "        strain_order.extend(new_list2)\n",
    "\n",
    "for phylon in phylon_order: # must be done after the first loop\n",
    "    if 'unchar' in phylon:\n",
    "        phylon_aff_binarized_single = A_binarized.loc[phylon, single_phylon_strains]\n",
    "        phylon_aff_binarized_multi = A_binarized.loc[phylon, multi_phylon_strains]\n",
    "    \n",
    "        inPhylon_single = phylon_aff_binarized_single == 1\n",
    "        inPhylon_multi = phylon_aff_binarized_multi == 1\n",
    "    \n",
    "        list1 = phylon_aff_binarized_single[inPhylon_single].index.tolist()\n",
    "        list2 = phylon_aff_binarized_multi[inPhylon_multi].index.tolist()\n",
    "        new_list1 = list(set(list1) - set(strain_order)) # ensures no double-counting\n",
    "        new_list2 = list(set(list2) - set(strain_order)) # ensures no double-counting\n",
    "        \n",
    "        strain_order.extend(new_list1)\n",
    "        strain_order.extend(new_list2)\n",
    "\n",
    "strain_order += unchar_strain_order\n",
    "\n",
    "# A-binarized\n",
    "sns.clustermap(A_binarized.loc[phylon_order, strain_order], cmap='Greys', xticklabels=False, row_cluster=False, col_cluster=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff678c0-445a-426b-9bc2-a799638dd7f7",
   "metadata": {},
   "source": [
    "# Number of rare genes per strain per phylon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b694a3a6-0f32-409f-8222-d89e7f276ba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = [\n",
    "    # Shades of red/orange/yellow\n",
    "    \"Red\",\n",
    "    \"IndianRed\",\n",
    "    \"DarkRed\",\n",
    "    \"FireBrick\",\n",
    "    \"Tomato\",\n",
    "    \"Gold\",\n",
    "    \"DarkGoldenrod\",\n",
    "    \"Goldenrod\",\n",
    "    # Other species\n",
    "    \"Green\",\n",
    "    \"Blue\",\n",
    "    \"Purple\",\n",
    "    \"Cyan\",\n",
    "    \"Magenta\",\n",
    "    \"Lime\",\n",
    "    \"Pink\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b5dd7e-1f9a-4dea-b21b-46438759e7d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_strains(phylon, A_binarized = A_binarized):\n",
    "    phylon_membership = A_binarized.loc[phylon]\n",
    "    return (phylon_membership[phylon_membership == 1]).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52088d76-9214-481b-8438-007597b0fa5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e5b9d4-56e8-45ff-bb29-c8dc99033979",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = []\n",
    "for phylon in characterized_order:\n",
    "    strains = get_strains(phylon)  # Your function to get strains for the phylon\n",
    "    rare_gene_counts = df_rare[strains].sum()  # Summing rare gene counts for each strain\n",
    "    for strain, count in rare_gene_counts.items():\n",
    "        data.append({'Phylon': phylon, 'Strain': strain, 'RareGeneCount': count})\n",
    "\n",
    "# Convert to a dataframe\n",
    "plot_data = pd.DataFrame(data)\n",
    "\n",
    "# Create the boxenplot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.boxenplot(data=plot_data, x='Phylon', y='RareGeneCount', palette=custom_colors)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Phylon')\n",
    "plt.ylabel('Rare Genes per Strain')\n",
    "plt.title('Distribution of Rare Genes per Strain in Each Phylon')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e52c4162-5083-45be-b817-4cd3cb1a0bb9",
   "metadata": {},
   "source": [
    "# AMR rare genes across phylons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d3ef3e-8c3d-4a6c-aac6-3f199a69b7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "amr = pd.read_csv('../../data/processed/amrfinder/output', sep = '\\t')\n",
    "amr['Protein identifier'] = amr['Protein identifier'].apply(lambda x: x.split('A')[0])\n",
    "amr = amr.sort_values('% Coverage of reference sequence')\n",
    "amr = amr.drop_duplicates(subset='Protein identifier', keep=\"last\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f95b52-183a-4f6e-ba3a-1c4b1da99ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_amr_genes = [x for x in amr['Protein identifier'] if x in df_rare.index]\n",
    "\n",
    "amr_presence = pd.DataFrame(np.zeros((len(rare_amr_genes), len(characterized_order))), index=rare_amr_genes, columns=characterized_order)\n",
    "amr_counts = pd.DataFrame(np.zeros((len(rare_amr_genes), len(characterized_order))), index=rare_amr_genes, columns=characterized_order)\n",
    "for phylon in characterized_order:\n",
    "    strains = get_strains(phylon)\n",
    "    amr_presence[phylon] = df_rare.loc[rare_amr_genes, strains].sum(axis=1)/len(strains)\n",
    "    amr_counts[phylon] = df_rare.loc[rare_amr_genes, strains].sum(axis=1)\n",
    "\n",
    "data = []\n",
    "for phylon in characterized_order:\n",
    "    strains = get_strains(phylon)  # Your function to get strains for the phylon\n",
    "    rare_gene_counts = df_rare.loc[rare_amr_genes, strains].sum()  # Summing rare gene counts for each strain\n",
    "    for strain, count in rare_gene_counts.items():\n",
    "        data.append({'Phylon': phylon, 'Strain': strain, 'RareGeneCount': count})\n",
    "plot_data = pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42100afd-7c78-4147-8350-d607fad9962a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_complete.loc[([x for x in amr['Protein identifier'] if x in df_genes_complete.index])].sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e261e7-aaa6-415b-951a-7efe54c64639",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_complete.loc[(rare_amr_genes)].sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26025041-e766-4dad-9933-76be02939bc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "# Assuming `amr_presence` is already computed\n",
    "row_sums = amr_counts.sum(axis=1)  # Sum per row (values per rare AMR gene)\n",
    "col_sums = amr_presence.sum(axis=0)  # Sum per column (values per phylon)\n",
    "\n",
    "# Retrieve Subclass values for the rare AMR genes\n",
    "subclass_values = amr.set_index('Protein identifier').loc[rare_amr_genes].Class\n",
    "\n",
    "# Sort the rare AMR genes by their subclass\n",
    "sorted_rare_amr_genes = subclass_values.sort_values().index\n",
    "\n",
    "# Recompute row sums for the sorted rare AMR genes\n",
    "sorted_row_sums = row_sums[sorted_rare_amr_genes]\n",
    "\n",
    "# Set up the figure and gridspec\n",
    "fig = plt.figure(figsize=(12, 12))\n",
    "gs = gridspec.GridSpec(2, 2, width_ratios=[1, 0.2], height_ratios=[0.2, 1], wspace=0.05, hspace=0.05)\n",
    "\n",
    "# Heatmap\n",
    "ax_heatmap = plt.subplot(gs[1, 0])\n",
    "sns.heatmap(\n",
    "    amr_presence.loc[sorted_rare_amr_genes],\n",
    "    cmap='Greys',\n",
    "    cbar=False,  # Disable default colorbar\n",
    "    ax=ax_heatmap\n",
    ")\n",
    "ax_heatmap.set_xlabel('Phylon')\n",
    "ax_heatmap.set_ylabel('Rare AMR Genes Percentage Presence')\n",
    "ax_heatmap.set_yticks([])  # Remove x-ticks\n",
    "# ax_heatmap.set_xticks([])\n",
    "# # Add horizontal lines to separate subclasses\n",
    "# subclass_boundaries = subclass_values.loc[sorted_rare_amr_genes].values\n",
    "# for i in range(1, len(subclass_boundaries)):\n",
    "#     if subclass_boundaries[i] != subclass_boundaries[i - 1]:\n",
    "#         ax_heatmap.axhline(i, color='black', linewidth=.5, alpha =.5, linestyle='-')\n",
    "\n",
    "# Top boxplot (column sums) aligned with heatmap\n",
    "ax_top_barplot = plt.subplot(gs[0, 0])\n",
    "sns.boxenplot(data=plot_data, x='Phylon', y='RareGeneCount', palette=custom_colors, ax=ax_top_barplot)\n",
    "ax_top_barplot.set_xticks([])  # Remove x-ticks\n",
    "ax_top_barplot.set_ylabel('# AMR Genes')\n",
    "ax_top_barplot.set_title('Distribution of rare AMR Genes per strain in Phylon')\n",
    "ax_top_barplot.tick_params(axis='x', which='both', bottom=False)\n",
    "\n",
    "# Align the top barplot width with the heatmap\n",
    "pos_heatmap = ax_heatmap.get_position()\n",
    "pos_top_bar = ax_top_barplot.get_position()\n",
    "ax_top_barplot.set_position([pos_heatmap.x0, pos_top_bar.y0, pos_heatmap.width, pos_top_bar.height])\n",
    "\n",
    "# Side barplot (row sums)\n",
    "unique_subclasses = subclass_values.sort_values().unique()\n",
    "subclass_palette = sns.color_palette(\"tab20\", len(unique_subclasses))\n",
    "subclass_colors = subclass_values.map(dict(zip(unique_subclasses, subclass_palette)))\n",
    "\n",
    "ax_side_barplot = plt.subplot(gs[1, 1])\n",
    "sns.barplot(\n",
    "    y=amr_presence.loc[sorted_rare_amr_genes].index,\n",
    "    x=row_sums.loc[sorted_rare_amr_genes],\n",
    "    ax=ax_side_barplot,\n",
    "    orient='h',\n",
    "    dodge=False,  # Prevent spacing between bars\n",
    "    palette=subclass_colors.loc[sorted_rare_amr_genes].values,\n",
    "    width=1\n",
    ")\n",
    "# ax_side_barplot.set_xticks([])  # Remove x-ticks\n",
    "ax_side_barplot.set_yticks([])  # Remove y-ticks\n",
    "ax_side_barplot.set_xlabel('Number of Strains')\n",
    "ax_side_barplot.set_ylabel('Genes')\n",
    "ax_side_barplot.tick_params(axis='y', which='both', left=False)\n",
    "\n",
    "# Add the colorbar manually\n",
    "cbar_ax = fig.add_axes([pos_heatmap.x1 - 0.73, pos_heatmap.y0, 0.02, pos_heatmap.height])\n",
    "sm = plt.cm.ScalarMappable(cmap='Greys', norm=plt.Normalize(vmin=amr_presence.min().min(), vmax=amr_presence.max().max()))\n",
    "cbar = fig.colorbar(sm, cax=cbar_ax)\n",
    "\n",
    "# Create the legend for the subclass colors\n",
    "handles = []\n",
    "for subclass, color in zip(unique_subclasses, subclass_palette):\n",
    "    handles.append(plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=color, markersize=10, label=subclass))\n",
    "\n",
    "ax_side_barplot.legend(handles=handles, title='Antibiotic Type', bbox_to_anchor=(1.1, 1), loc='upper left', fontsize=12, ncols=1)\n",
    "\n",
    "phylon_handles = []\n",
    "for phylon, color in zip(characterized_order, custom_colors):\n",
    "    phylon_handles.append(\n",
    "        plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=color, markersize=10, label=phylon)\n",
    "    )\n",
    "\n",
    "# Add the phylon legend to the top boxplot\n",
    "ax_top_barplot.legend(\n",
    "    handles=phylon_handles, \n",
    "    title='Phylon', \n",
    "    bbox_to_anchor=(1.05, 1),  # Position legend to the right of the top boxplot\n",
    "    loc='upper left', \n",
    "    fontsize=10, \n",
    "    ncol=3\n",
    ")\n",
    "\n",
    "plt.savefig('amr_presence_plot.svg', format='svg')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7797d379-55b2-434c-9f1f-4288a5cfc4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "# Assuming `amr_presence` is already computed\n",
    "row_sums = amr_counts.sum(axis=1)  # Sum per row (values per rare AMR gene)\n",
    "col_sums = amr_presence.sum(axis=0)  # Sum per column (values per phylon)\n",
    "\n",
    "# Retrieve Subclass values for the rare AMR genes\n",
    "subclass_values = amr.set_index('Protein identifier').loc[rare_amr_genes].Class\n",
    "\n",
    "# Sort the rare AMR genes by their subclass\n",
    "sorted_rare_amr_genes = subclass_values.sort_values().index\n",
    "\n",
    "# Recompute row sums for the sorted rare AMR genes\n",
    "sorted_row_sums = row_sums[sorted_rare_amr_genes]\n",
    "\n",
    "# Set up the figure and gridspec\n",
    "fig = plt.figure(figsize=(12, 12))\n",
    "gs = gridspec.GridSpec(2, 2, width_ratios=[1, 0.2], height_ratios=[0.075, 1], wspace=0.05, hspace=0.05)\n",
    "\n",
    "# Heatmap\n",
    "ax_heatmap = plt.subplot(gs[1, 0])\n",
    "sns.heatmap(\n",
    "    amr_presence.loc[sorted_rare_amr_genes],\n",
    "    cmap='Greys',\n",
    "    cbar=False,  # Disable default colorbar\n",
    "    ax=ax_heatmap\n",
    ")\n",
    "ax_heatmap.set_xlabel('Phylon')\n",
    "ax_heatmap.set_ylabel('Rare AMR Genes Percentage Presence')\n",
    "ax_heatmap.set_yticks([])  # Remove y-ticks\n",
    "ax_heatmap.set_xticks([])  # Remove x-ticks\n",
    "\n",
    "# Add blocks of color representing the phylons (adjust height to 1/3)\n",
    "phylon_colors = dict(zip(characterized_order, custom_colors))\n",
    "\n",
    "# Plot colored blocks instead of the top barplot (adjust height to 1/3)\n",
    "ax_top_color_blocks = plt.subplot(gs[0, 0])\n",
    "\n",
    "block_height = 1 / 3  # Set block height to 1/3 of the original size\n",
    "\n",
    "# Loop through phylon list to create color blocks\n",
    "for i, phylon in enumerate(characterized_order):\n",
    "    ax_top_color_blocks.add_patch(plt.Rectangle(\n",
    "        (i, 0), 1, block_height, color=phylon_colors[phylon], lw=0))  # Add blocks with phylon colors\n",
    "\n",
    "# Adjust the limits of the plot to fit the smaller blocks\n",
    "ax_top_color_blocks.set_xlim(0, len(characterized_order))  # Adjust x limits to fit blocks\n",
    "ax_top_color_blocks.set_ylim(0, block_height)  # Set y-limits to fit smaller block height\n",
    "ax_top_color_blocks.set_xticks([])  # Remove x-ticks\n",
    "ax_top_color_blocks.set_yticks([])  # Remove y-ticks\n",
    "ax_top_color_blocks.set_title('Phylon Representation')\n",
    "\n",
    "# Side barplot (row sums)\n",
    "unique_subclasses = subclass_values.sort_values().unique()\n",
    "subclass_palette = sns.color_palette(\"tab20\", len(unique_subclasses))\n",
    "subclass_colors = subclass_values.map(dict(zip(unique_subclasses, subclass_palette)))\n",
    "\n",
    "ax_side_barplot = plt.subplot(gs[1, 1])\n",
    "sns.barplot(\n",
    "    y=amr_presence.loc[sorted_rare_amr_genes].index,\n",
    "    x=row_sums.loc[sorted_rare_amr_genes],\n",
    "    ax=ax_side_barplot,\n",
    "    orient='h',\n",
    "    dodge=False,  # Prevent spacing between bars\n",
    "    palette=subclass_colors.loc[sorted_rare_amr_genes].values,\n",
    "    width=1\n",
    ")\n",
    "ax_side_barplot.set_yticks([])  # Remove y-ticks\n",
    "ax_side_barplot.set_xlabel('Number of Strains')\n",
    "ax_side_barplot.set_ylabel('Genes')\n",
    "ax_side_barplot.tick_params(axis='y', which='both', left=False)\n",
    "\n",
    "# Add the colorbar manually\n",
    "cbar_ax = fig.add_axes([ax_heatmap.get_position().x1 - 0.73, ax_heatmap.get_position().y0, 0.02, ax_heatmap.get_position().height])\n",
    "sm = plt.cm.ScalarMappable(cmap='Greys', norm=plt.Normalize(vmin=amr_presence.min().min(), vmax=amr_presence.max().max()))\n",
    "cbar = fig.colorbar(sm, cax=cbar_ax)\n",
    "\n",
    "# Create the legend for the subclass colors\n",
    "handles = []\n",
    "for subclass, color in zip(unique_subclasses, subclass_palette):\n",
    "    handles.append(plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=color, markersize=10, label=subclass))\n",
    "\n",
    "ax_side_barplot.legend(handles=handles, title='Antibiotic Type', bbox_to_anchor=(1.1, 1), loc='upper left', fontsize=12, ncols=1)\n",
    "\n",
    "# Create the legend for the phylons\n",
    "phylon_handles = []\n",
    "for phylon, color in zip(characterized_order, custom_colors):\n",
    "    phylon_handles.append(\n",
    "        plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=color, markersize=10, label=phylon)\n",
    "    )\n",
    "\n",
    "# Add the phylon legend to the top color blocks\n",
    "ax_top_color_blocks.legend(\n",
    "    handles=phylon_handles, \n",
    "    title='Phylon', \n",
    "    bbox_to_anchor=(1.05, 1.9),  # Position legend to the right of the top color blocks\n",
    "    loc='upper left', \n",
    "    fontsize=10, \n",
    "    ncol=3\n",
    ")\n",
    "\n",
    "plt.savefig('amr_presence_plot_with_smaller_color_blocks_height.svg', format='svg')\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pan-phylon]",
   "language": "python",
   "name": "conda-env-pan-phylon-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
