{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcce15b1-efc4-4803-ba41-d7b99d2bd2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Base imports\n",
    "import os\n",
    "import pickle\n",
    "import re\n",
    "\n",
    "# Compute imports\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from tqdm.notebook import tqdm, trange\n",
    "\n",
    "# Plotting imports\n",
    "import matplotlib\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotly import express as px\n",
    "\n",
    "# ML import\n",
    "from sklearn.decomposition import NMF\n",
    "from sklearn.metrics import mean_squared_error, median_absolute_error\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from scipy import sparse\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['svg.fonttype'] = 'none'\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Arial'\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "sns.set_style('ticks')\n",
    "matplotlib.rcParams['text.color'] = '#000000'\n",
    "matplotlib.rcParams['axes.labelcolor'] = '#000000'\n",
    "matplotlib.rcParams['xtick.color'] = '#000000'\n",
    "matplotlib.rcParams['ytick.color'] = '#000000'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "440eb2fa-c4f1-418a-ad9b-eb5b9384579f",
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_DATA = '../../data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b639c502-440d-44a5-bfc7-10381e784b72",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "DF_GENES = '../../data/processed/cd-hit-results/sim80/Ebacter_strain_by_gene.pickle.gz'\n",
    "ENRICHED_METADATA = '../../data/metadata/enriched_metadata.csv'\n",
    "DF_EGGNOG = '../../data/processed/df_eggnog.csv'\n",
    "\n",
    "DF_CORE_COMPLETE = '../../data/processed/CAR_genomes/df_core_complete.pickle'\n",
    "DF_ACC_COMPLETE = '../../data/processed/CAR_genomes/df_acc_complete.pickle'\n",
    "DF_RARE_COMPLETE = '../../data/processed/CAR_genomes/df_rare_complete.pickle'\n",
    "\n",
    "L_BINARIZED = '../../data/processed/nmf-outputs/L_binarized.csv'\n",
    "A_BINARIZED = '../../data/processed/nmf-outputs/A_binarized.csv'\n",
    "L_MATRIX = '../../data/processed/nmf-outputs/L.csv'\n",
    "A_MATRIX = '../../data/processed/nmf-outputs/A.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a5aa95-16dd-43aa-a4a5-40e0ee14350f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in A_binarized matrix\n",
    "A_binarized = pd.read_csv(A_BINARIZED, index_col=0)\n",
    "L_binarized = pd.read_csv(L_BINARIZED, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d617cf-ad31-4f87-bc82-9d9616823d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rare = pd.read_pickle(DF_RARE_COMPLETE)\n",
    "df_acc = pd.read_pickle(DF_ACC_COMPLETE)\n",
    "df_core = pd.read_pickle(DF_CORE_COMPLETE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1380e7-b85c-4fd8-966a-6c77f536dd07",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(ENRICHED_METADATA, index_col=0, dtype='object')\n",
    "\n",
    "display( metadata.shape, metadata.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42770db-2ec2-4342-8cb6-638d4ccdb4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in (full) P matrix\n",
    "df_genes = pd.read_pickle(DF_GENES)\n",
    "\n",
    "# Filter metadata for Complete sequences only\n",
    "metadata_complete = metadata[metadata.genome_status == 'Complete'] # filter for only Complete sequences\n",
    "\n",
    "# Filter P matrix for Complete sequences only\n",
    "df_genes_complete = df_genes[metadata_complete.genome_id].copy()\n",
    "df_genes_complete.fillna(0, inplace=True) # replace N/A with 0\n",
    "df_genes_complete = df_genes_complete.sparse.to_dense().astype('int8') # densify & typecast to int8 for space and compute reasons\n",
    "inCompleteseqs = df_genes_complete.sum(axis=1) > 0 # filter for genes found in complete sequences\n",
    "df_genes_complete = df_genes_complete[inCompleteseqs]\n",
    "\n",
    "df_genes_complete.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49554230-8993-4d93-a31e-7e72d285472d",
   "metadata": {},
   "outputs": [],
   "source": [
    "phylon_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'unchar-1',\n",
    " 'unchar-2',\n",
    " 'unchar-3',\n",
    " 'unchar-4',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']\n",
    "\n",
    "characterized_order = ['hormaechei-xiangfangensis',\n",
    " 'hormaechei-oharae',\n",
    " 'hormaechei-steigerwaltii-2',\n",
    " 'hormaechei-steigerwaltii-1',\n",
    " 'hormaechei-steigerwaltii-3',\n",
    " 'hormaechei-hormaechei',\n",
    " 'hormaechei-hoffmannii-1',\n",
    " 'hormaechei-hoffmannii-2',\n",
    " 'roggenkampii',\n",
    " 'asburiae',\n",
    " 'kobei',\n",
    " 'bugandensis',\n",
    " 'cancerogenous',\n",
    " 'ludwigii',\n",
    " 'cloacae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e380d66-f5a9-4edb-9316-194158967539",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in eggNOG annotations\n",
    "df_eggnog = pd.read_csv(DF_EGGNOG, index_col=0)\n",
    "df_eggnog.fillna('-', inplace=True)\n",
    "df_eggnog.COG_category = df_eggnog.COG_category.apply(lambda x: x[0])\n",
    "display(\n",
    "    df_eggnog.shape,\n",
    "    df_eggnog.head()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27cd769b-0504-4db8-9485-4a92ca1f6d0d",
   "metadata": {},
   "source": [
    "## Analysis of Rare Genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9edcb26f-4b32-4ca5-8cde-18ee7d448620",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_genes = df_rare.index\n",
    "rare_geneome_eggnog = df_eggnog.loc[rare_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb24e743-481d-4038-8b7f-1f987d7a670f",
   "metadata": {},
   "outputs": [],
   "source": [
    "BAKTA_ANNOTATIONS = '../../data/processed/bakta_gene_annotations.csv'\n",
    "bakta_annotations = pd.read_csv(BAKTA_ANNOTATIONS, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce510f7-b5af-4320-b92b-12a1ec2f1443",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional: drop all genes which are predicted as hypothetical \n",
    "df_rare_annot = df_rare\n",
    "# to_drop = bakta_annotations[(bakta_annotations.Product.str.contains('ypothetical'))].index\n",
    "# to_drop = [x for x in to_drop if x in df_rare.index]\n",
    "# df_rare_annot = df_rare.drop(to_drop).astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801dca0e-b34d-4f7f-b2a1-065453a847fd",
   "metadata": {},
   "source": [
    "# Extract gene locations for determination of gene proximity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c07d0bb-89bf-4f09-8b60-935d1da9302b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_files = '../../data/processed/bakta/'\n",
    "header_to_allele = '../../data/processed/cd-hit-results/header_to_allele_80.pickle.gz'\n",
    "cd_hit_headers = '../../data/processed/cd-hit-results/rep_headers.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9bb3870-d340-47e8-aabe-83c03762e740",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h2a = pd.read_pickle(header_to_allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12c8be12-fc37-424e-8723-8e12fdb2270e",
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = open(cd_hit_headers).readlines()\n",
    "headers = [x[1:13] for x in headers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc8fc51-9cb8-4145-a0cc-37d7048ab535",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene to alleles within\n",
    "cluster_to_alleles = {}\n",
    "\n",
    "# Iterate over the original dictionary\n",
    "for allele, cluster in tqdm(df_h2a.items()):\n",
    "    cluster = cluster.split('A')[0]\n",
    "    # If the cluster is not yet in the new dictionary, add it with an empty list\n",
    "    if cluster not in cluster_to_alleles:\n",
    "        cluster_to_alleles[cluster] = []\n",
    "    # Append the allele to the list of alleles for this cluster\n",
    "    cluster_to_alleles[cluster].append(allele)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5c2559-39d0-4aad-afce-7ba40985b6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_to_tag = {}\n",
    "for genome in tqdm(os.listdir(bakta_files)):\n",
    "    file = open(bakta_files + genome + '/' + genome + '.gff3')\n",
    "    file.seek(0)\n",
    "    text = file.read(10000)\n",
    "    loc = text.find('locus_tag=')\n",
    "    tag = text[loc+10:loc+16]\n",
    "    genome_to_tag[genome] = tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011aeb39-8110-4929-b101-b9ec4e5cb396",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_attr(attributes, attr_id, ignore=False):\n",
    "    \"\"\"\n",
    "    Helper function for parsing GFF annotations\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    attributes : str\n",
    "        Attribute string\n",
    "    attr_id : str\n",
    "        Attribute ID\n",
    "    ignore : bool\n",
    "        If true, ignore errors if ID is not in attributes (default: False)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str, optional\n",
    "        Value of attribute\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        return re.search(attr_id + \"=(.*?)(;|$)\", attributes).group(1)\n",
    "    except AttributeError:\n",
    "        if ignore:\n",
    "            return None\n",
    "        else:\n",
    "            raise ValueError(\"{} not in attributes: {}\".format(attr_id, attributes))\n",
    "\n",
    "def gff2pandas(gff_file, feature=[\"CDS\"], index=None):\n",
    "    \"\"\"\n",
    "    Converts GFF file(s) to a Pandas DataFrame\n",
    "    Parameters\n",
    "    ----------\n",
    "    gff_file : str or list\n",
    "        Path(s) to GFF file\n",
    "    feature: str or list\n",
    "        Name(s) of features to keep (default = \"CDS\")\n",
    "    index : str, optional\n",
    "        Column or attribute to use as index\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    df_gff: ~pandas.DataFrame\n",
    "        GFF formatted as a DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    # Argument checking\n",
    "    if isinstance(gff_file, str):\n",
    "        gff_file = [gff_file]\n",
    "\n",
    "    if isinstance(feature, str):\n",
    "        feature = [feature]\n",
    "\n",
    "    result = []\n",
    "\n",
    "    for gff in gff_file:\n",
    "        with open(gff, \"r\") as f:\n",
    "            lines = f.readlines()\n",
    "\n",
    "        # Get lines to skip\n",
    "        skiprow = [i for i, line in enumerate(lines) if line.startswith(\"#\")]\n",
    "       \n",
    "        # Read GFF\n",
    "        names = [\n",
    "            \"accession\",\n",
    "            \"source\",\n",
    "            \"feature\",\n",
    "            \"start\",\n",
    "            \"end\",\n",
    "            \"score\",\n",
    "            \"strand\",\n",
    "            \"phase\",\n",
    "            \"attributes\",\n",
    "        ]\n",
    "        DF_gff = pd.read_csv(gff, sep=\"\\t\", skiprows=skiprow, names=names, header=None, low_memory=False)\n",
    "        \n",
    "        region = DF_gff[DF_gff.feature == 'region']\n",
    "        region_len = int(region.iloc[0].end)\n",
    "\n",
    "        # Filter for CDSs\n",
    "        DF_cds = DF_gff[DF_gff.feature.isin(feature)]\n",
    "\n",
    "        # Sort by start position\n",
    "        # DF_cds = DF_cds.sort_values(\"start\")\n",
    "\n",
    "        DF_cds = DF_cds.copy() # get rid of copy warning\n",
    "        \n",
    "        # Extract attribute information\n",
    "        DF_cds[\"locus_tag\"] = DF_cds.attributes.apply(_get_attr, attr_id=\"locus_tag\")\n",
    "\n",
    "        result.append(DF_cds)\n",
    "        \n",
    "    DF_gff = pd.concat(result)\n",
    "    \n",
    "    if index:\n",
    "        if DF_gff[index].duplicated().any():\n",
    "            logging.warning(\"Duplicate {} detected. Dropping duplicates.\".format(index))\n",
    "            DF_gff = DF_gff.drop_duplicates(index)\n",
    "        DF_gff.set_index(\"locus_tag\", drop=True, inplace=True)\n",
    "\n",
    "    return DF_gff[['accession', 'start', 'end', 'locus_tag', 'strand']], region_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "938b4a53-473a-4992-bdf3-37cd6a94312f",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_vectors = {}\n",
    "def h2a(x):\n",
    "    try:\n",
    "        return df_h2a[x].split('A')[0]\n",
    "    except:\n",
    "        return None\n",
    "        \n",
    "for strain in tqdm(metadata_complete.genome_id):\n",
    "    DF_gff, size = gff2pandas(f'{PATH_TO_DATA}/processed/bakta/{strain}/{strain}.gff3')\n",
    "    DF_gff['gene'] = DF_gff.locus_tag.apply(lambda x: h2a(x))\n",
    "    # DF_gff = DF_gff[DF_gff.accession == DF_gff.accession.value_counts().index[0]]\n",
    "    # DF_gff = DF_gff[['gene','start', 'end', 'strand']]\n",
    "    # gene_order = (DF_gff.sort_values('start').gene.to_list())\n",
    " \n",
    "    strain_vectors[strain] = DF_gff.sort_values(by=['accession', 'start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc5925c-9649-454a-bb47-a2758512c7ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rare_genes = set(df_rare.index)\n",
    "\n",
    "strain_contigs = {}\n",
    "\n",
    "rare_genome_stats = pd.DataFrame(index = list(rare_genes), columns = ['sizes', 'genomes', 'all_neighbors'])\n",
    "for gene in rare_genes:\n",
    "    rare_genome_stats.loc[gene] = [[], [], set()]\n",
    "\n",
    "strain_contig_sizes = pd.Series({strain:[] for strain in metadata_complete.genome_id})\n",
    "\n",
    "for strain, df in tqdm(strain_vectors.items()):\n",
    "    strain_contigs[strain] = []\n",
    "    curr_contig = []\n",
    "    curr_accession = \"\"\n",
    "    for i,row in df.iterrows():\n",
    "        if curr_accession == row.accession or curr_accession == \"\":\n",
    "            if row.gene in rare_genes:\n",
    "\n",
    "                curr_contig.append(row.gene)\n",
    "\n",
    "            elif len(curr_contig) > 0:\n",
    "                strain_contigs[strain].append(curr_contig)\n",
    "                strain_contig_sizes.loc[strain].append(len(curr_contig))\n",
    "                for gene in curr_contig:\n",
    "                    rare_genome_stats.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "                    rare_genome_stats.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs[strain])-1))\n",
    "                    rare_genome_stats.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))  \n",
    "        \n",
    "                curr_contig = []\n",
    "        elif len(curr_contig) > 0:\n",
    "            strain_contigs[strain].append(curr_contig)\n",
    "            strain_contig_sizes.loc[strain].append(len(curr_contig))\n",
    "            for gene in curr_contig:\n",
    "                rare_genome_stats.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "                rare_genome_stats.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs[strain])-1))\n",
    "                rare_genome_stats.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))\n",
    "                                           \n",
    "            curr_contig = []\n",
    "        if row.gene in rare_genes and row.gene not in curr_contig:\n",
    "            curr_contig.append(row.gene)\n",
    "            \n",
    "        curr_accession = row.accession\n",
    "\n",
    "    if len(curr_contig) > 0:\n",
    "        strain_contigs[strain].append(curr_contig)\n",
    "        strain_contig_sizes.loc[strain].append(len(curr_contig))\n",
    "        for gene in curr_contig:\n",
    "            rare_genome_stats.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "            rare_genome_stats.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs[strain])-1))\n",
    "            rare_genome_stats.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))\n",
    "                \n",
    "        curr_contig = []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70263773-ab8b-48b2-8dc3-0c97fa3b34cb",
   "metadata": {},
   "source": [
    "# Repeated for only main chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "272ed2b3-d73b-4302-8f93-d566fc9c19bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_vectors_chrom = {}\n",
    "def h2a(x):\n",
    "    try:\n",
    "        return df_h2a[x].split('A')[0]\n",
    "    except:\n",
    "        return None\n",
    "        \n",
    "for strain in tqdm(metadata_complete.genome_id):\n",
    "    DF_gff, size = gff2pandas(f'{PATH_TO_DATA}/processed/bakta/{strain}/{strain}.gff3')\n",
    "    DF_gff['gene'] = DF_gff.locus_tag.apply(lambda x: h2a(x))\n",
    "    DF_gff = DF_gff[DF_gff.accession == DF_gff.accession.value_counts().index[0]]\n",
    "    DF_gff = DF_gff[['accession','gene','start', 'end', 'strand']]\n",
    "    # gene_order = (DF_gff.sort_values('start').gene.to_list())\n",
    " \n",
    "    strain_vectors_chrom[strain] = DF_gff.sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc3a7fb-a6f9-4524-9b6c-8be590d0713b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rare_genes_chrom = set(df_rare.index)\n",
    "\n",
    "strain_contigs_chrom = {}\n",
    "\n",
    "rare_genome_stats_chrom = pd.DataFrame(index = list(rare_genes_chrom), columns = ['sizes', 'genomes', 'all_neighbors'])\n",
    "for gene in rare_genes_chrom:\n",
    "    rare_genome_stats_chrom.loc[gene] = [[], [], set()]\n",
    "\n",
    "strain_contig_sizes_chrom = pd.Series({strain:[] for strain in metadata_complete.genome_id})\n",
    "\n",
    "for strain, df in tqdm(strain_vectors_chrom.items()):\n",
    "    if strain == '550.624':\n",
    "        continue\n",
    "    \n",
    "    strain_contigs_chrom[strain] = []\n",
    "    curr_contig = []\n",
    "    curr_accession = \"\"\n",
    "    for i,row in df.iterrows():\n",
    "        if curr_accession == row.accession or curr_accession == \"\":\n",
    "            if row.gene in rare_genes_chrom:\n",
    "\n",
    "                curr_contig.append(row.gene)\n",
    "\n",
    "            elif len(curr_contig) > 0:\n",
    "                strain_contigs_chrom[strain].append(curr_contig)\n",
    "                strain_contig_sizes_chrom.loc[strain].append(len(curr_contig))\n",
    "                for gene in curr_contig:\n",
    "                    rare_genome_stats_chrom.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "                    rare_genome_stats_chrom.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs_chrom[strain])-1))\n",
    "                    rare_genome_stats_chrom.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))  \n",
    "        \n",
    "                curr_contig = []\n",
    "        elif len(curr_contig) > 0:\n",
    "            strain_contigs_chrom[strain].append(curr_contig)\n",
    "            strain_contig_sizes_chrom.loc[strain].append(len(curr_contig))\n",
    "            for gene in curr_contig:\n",
    "                rare_genome_stats_chrom.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "                rare_genome_stats_chrom.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs_chrom[strain])-1))\n",
    "                rare_genome_stats_chrom.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))\n",
    "                                           \n",
    "            curr_contig = []\n",
    "        if row.gene in rare_genes_chrom and row.gene not in curr_contig:\n",
    "            curr_contig.append(row.gene)\n",
    "            \n",
    "        curr_accession = row.accession\n",
    "\n",
    "    if len(curr_contig) > 0:\n",
    "        strain_contigs_chrom[strain].append(curr_contig)\n",
    "        strain_contig_sizes_chrom.loc[strain].append(len(curr_contig))\n",
    "        for gene in curr_contig:\n",
    "            rare_genome_stats_chrom.loc[gene, 'sizes'].append(len(curr_contig))\n",
    "            rare_genome_stats_chrom.loc[gene, 'genomes'].append(strain + '_' + str(len(strain_contigs_chrom[strain])-1))\n",
    "            rare_genome_stats_chrom.loc[gene, 'all_neighbors'].update(set(curr_contig) - set([gene]))\n",
    "                \n",
    "        curr_contig = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0921fd3-dfad-4cd8-a623-96865d86868e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# analyze if singleton rare genes are usually hypothetical proteins or have other enrichments\n",
    "rare_genome_stats_chrom = rare_genome_stats_chrom.drop(rare_genome_stats_chrom[rare_genome_stats_chrom.sizes.apply(lambda x: len(x) == 0)].index)\n",
    "strain_contig_sizes_chrom = strain_contig_sizes_chrom.drop('550.624')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12cbce72-daca-4bd8-b944-d7653892c367",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_genome_stats_chrom.sort_values('sizes').sizes.apply(lambda x: len(x)).hist()\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54cc47ec-5c52-4024-a465-f221054f5b5b",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eb94225-2f55-4bda-807a-ce0d1b6d39e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Read the data (alleles)\n",
    "P_allele = pd.read_pickle('../../data/processed/cd-hit-results/sim80/Ebacter_strain_by_allele.pickle.gz')\n",
    "P_allele = P_allele.loc[:, df_core.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29178617-d087-4bca-b078-956d0ecb25db",
   "metadata": {},
   "outputs": [],
   "source": [
    "P_allele = P_allele.loc[:, df_core.columns].fillna(0)\n",
    "mask = np.any(P_allele.values == 1, axis = 1)\n",
    "P_allele = P_allele[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8962fd-5460-48e4-8280-04b61bb11ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "relevant_alleles = [x for x in P_allele.index if x.split('A')[0] in df_rare.index]\n",
    "len(relevant_alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "476187ce-0c0e-4df9-b0c7-fedd95705042",
   "metadata": {},
   "outputs": [],
   "source": [
    "P_allele = P_allele.loc[relevant_alleles].fillna(0).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bd702f9-ed4e-4fe1-b801-315456d2fc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the FASTA file\n",
    "from Bio import SeqIO\n",
    "\n",
    "fasta_file = \"../../data/processed/cd-hit-results/sim80/Ebacter_nr.faa\"\n",
    "\n",
    "\n",
    "gene_lengths = pd.DataFrame(index = df_rare.index, columns = ['lens', 'median_len'])\n",
    "gene_lengths['lens'] = [[] for _ in range(len(df_rare))]\n",
    "\n",
    "# Iterate through the FASTA file and store sequences associated with headers in gene_list\n",
    "for record in tqdm(SeqIO.parse(fasta_file, \"fasta\")):\n",
    "    # print(record.id, len(record.seq))\n",
    "    if record.id in P_allele.index:\n",
    "        gene_lengths.loc[record.id.split('A')[0], 'lens'].append(len(record.seq))\n",
    "gene_lengths['median_len'] = (gene_lengths.lens).apply(lambda x: np.median(x))\n",
    "gene_lengths['cog'] = gene_lengths.apply(lambda x: df_eggnog.loc[x.name, 'COG_category'][0], axis=1)\n",
    "gene_lengths = gene_lengths.replace('-','S')\n",
    "gene_lengths['alleles'] = gene_lengths.lens.apply(lambda x: len(x))\n",
    "\n",
    "def get_cog_super(cog):\n",
    "    if cog in ['J', 'A', 'K', 'L']:\n",
    "        return 'INFORMATION STORAGE AND PROCESSING'\n",
    "    elif cog in 'D Y V T M N Z W U O'.split():\n",
    "        return 'CELLULAR PROCESSES AND SIGNALING'\n",
    "    elif cog in 'C G E F H I P Q'.split():\n",
    "        return 'METABOLISM'\n",
    "    else:\n",
    "        return 'POORLY CHARACTERIZED'\n",
    "\n",
    "gene_lengths['cog_super'] = gene_lengths.cog.apply(lambda x: get_cog_super(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "240e8777-abfe-4359-a1f5-c121b81ebebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_colors = [\n",
    "    # Shades of red/orange/yellow\n",
    "    \"Red\",\n",
    "    \"IndianRed\",\n",
    "    \"DarkRed\",\n",
    "    \"FireBrick\",\n",
    "    \"Tomato\",\n",
    "    \"Gold\",\n",
    "    \"DarkGoldenrod\",\n",
    "    \"Goldenrod\",\n",
    "    # Other species\n",
    "    \"Green\",\n",
    "    \"Blue\",\n",
    "    \"Purple\",\n",
    "    \"Cyan\",\n",
    "    \"Magenta\",\n",
    "    \"Lime\",\n",
    "    \"Pink\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b7e7a6-4fd2-4be8-bc9e-ee98dcc4640d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_strains(phylon, A_binarized = A_binarized):\n",
    "    phylon_membership = A_binarized.loc[phylon]\n",
    "    return (phylon_membership[phylon_membership == 1]).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7181e4a-b360-44b8-994b-e32a5b09de51",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "\n",
    "plot_data = pd.DataFrame(columns = ['Phylon', 'Genome', 'Region_Size'])\n",
    "df = strain_contig_sizes_chrom.explode().reset_index()\n",
    "df.columns = ['Genome', 'Region_Size']\n",
    "for phylon in characterized_order:\n",
    "    strains = get_strains(phylon)  # Your function to get strains for the phylon\n",
    "    contiguous_sizes = df[df['Genome'].isin(strains)]  # Summing rare gene counts for each strain\n",
    "    contiguous_sizes.Region_Size = contiguous_sizes.Region_Size.apply(lambda x: np.log2(x))\n",
    "    contiguous_sizes = contiguous_sizes.assign(Phylon=[phylon]*len(contiguous_sizes))\n",
    "    plot_data = pd.concat([plot_data,contiguous_sizes], ignore_index=True)\n",
    "\n",
    "# Create the boxenplot\n",
    "plt.figure(figsize=(6, 8))\n",
    "sns.boxenplot(data=plot_data, x='Phylon', y='Region_Size', palette=custom_colors)\n",
    "plt.xticks([],rotation=45, ha='right')\n",
    "plt.xlabel('Phylon')\n",
    "plt.ylabel('Log2 Rare Genes per Strain')\n",
    "plt.title('Distribution of Chromosomal Rare Gene \\n Region Sizes Across Strains in Each Phylon')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dad0632-fc28-4839-bac0-0fcbf19acefe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2f6ac83a-e669-43df-a62e-36f047faa748",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# TN Central"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be6d4c04-fcc4-4efc-9bc6-e2fe53926d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results_tncentral = pd.read_csv('../../data/blastdbs/tn_central_enrichment.txt', sep='\\t', header=None)\n",
    "blast_results_tncentral.columns = ['query', 'target', 'identity', 'len', 'mismatch', 'gapopen', 'qstart', 'qend', 'tstart', 'tend',\n",
    "                                     'eval', 'bitscore']\n",
    "blast_results_tncentral = blast_results_tncentral[blast_results_tncentral.identity > 80]\n",
    "blast_results_tncentral['query'] = blast_results_tncentral['query'].apply(lambda x: x.split('A')[0])\n",
    "blast_results_tncentral['target'] = blast_results_tncentral['target'].apply(lambda x: x.split('(')[0])\n",
    "blast_results_tncentral = blast_results_tncentral.sort_values(by=['query', 'eval'], ascending=[True, True])\n",
    "blast_results_tncentral  = blast_results_tncentral.drop_duplicates(subset='query', keep='first')\n",
    "blast_results_tncentral = blast_results_tncentral.set_index('query')\n",
    "blast_results_tncentral_rare = blast_results_tncentral.loc[[x for x in blast_results_tncentral.index if x in rare_genes]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f82ac6-fcaa-4f49-947b-9571a61f85cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_genome_stats_chrom.loc[[x for x in blast_results_tncentral_rare.index if x in rare_genome_stats_chrom.index]].sort_values('sizes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de6640e-c179-4b4b-97cf-b8f3db101186",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10fe223e-e2dd-4ba9-9c5f-8aa83da046f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rare.loc[[x for x in blast_results_tncentral.index if x in df_rare.index]].sum().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85befa6f-fe60-4edb-a3bc-84df331ba2fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_acc.loc[[x for x in blast_results_tncentral.index if x in df_acc.index]].sum().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2cd842c-b13d-40b7-baca-3404db9d0b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_complete.loc[[x for x in blast_results_tncentral.index if x in df_genes_complete.index]].sum().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "884efdbf-c17d-4eab-b97b-cc00880614ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "hypothetical_proteins = bakta_annotations.loc[rare_genome_stats_chrom.index][bakta_annotations.loc[rare_genome_stats_chrom.index].Product == 'hypothetical protein'].index\n",
    "others = [x for x in rare_genome_stats_chrom.index if x not in hypothetical_proteins]\n",
    "\n",
    "plt.violinplot(rare_genome_stats_chrom.loc[hypothetical_proteins].sizes.apply(lambda x: len(x)).sort_values())\n",
    "plt.violinplot(rare_genome_stats_chrom.loc[others].sizes.apply(lambda x: len(x)).sort_values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5b4b13-8816-4c80-954a-516994e58f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create a new DataFrame combining both groups\n",
    "df_combined = pd.DataFrame({\n",
    "    'Size': list(rare_genome_stats_chrom.loc[hypothetical_proteins].sizes.apply(lambda x: np.mean(x)).sort_values()) +\n",
    "            list(rare_genome_stats_chrom.loc[others].sizes.apply(lambda x: np.mean(x)).sort_values()),\n",
    "    'Group': ['Hypothetical Protein'] * len(hypothetical_proteins) +\n",
    "             ['Other'] * len(others)\n",
    "})\n",
    "\n",
    "# Create the violin plot\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.violinplot(x='Group', y='Size', data=df_combined)\n",
    "\n",
    "# Optional: Set plot labels\n",
    "plt.title('Violin Plot of Sizes for Hypothetical Proteins vs Others')\n",
    "plt.xlabel('Group')\n",
    "plt.ylabel('Size')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe8e621-5ee8-40a2-95aa-7013015f774a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_annotations.loc[rare_genome_stats.index].Product.value_counts().head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f77774cb-be97-41ab-8679-2746fd8dda53",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_annotations.loc[rare_genome_stats_chrom.index].Product.value_counts().head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ad65614-8968-4b27-9234-f72e3d108790",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b688e2e5-bfe4-4384-bb96-7da6bfdddc63",
   "metadata": {},
   "source": [
    "# AMR Genes and Virulence Factors seen to be neighbors in some genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03eefae5-ddcb-47e3-80f5-304bb7d2256a",
   "metadata": {},
   "outputs": [],
   "source": [
    "amr = pd.read_csv('../../data/processed/amrfinder/output', sep = '\\t')\n",
    "amr['Protein identifier'] = amr['Protein identifier'].apply(lambda x: x.split('A')[0])\n",
    "amr = amr.sort_values('% Coverage of reference sequence')\n",
    "amr = amr.drop_duplicates(subset='Protein identifier', keep=\"last\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1575528-7558-41bb-9508-b033561ad547",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import blast results and faa file to translate between IDs\n",
    "blast_results = pd.read_csv('../../data/blastdbs/vfdb.txt', sep='\\t', header=None)\n",
    "                           \n",
    "blast_results.columns = ['query', 'target', 'identity', 'len', 'mismatch', 'gapopen', 'qstart', 'qend', 'tstart', 'tend',\n",
    "                                     'eval', 'bitscore']\n",
    "blast_results = blast_results[blast_results.identity > 80]\n",
    "blast_results['query'] = blast_results['query'].apply(lambda x: x.split('A')[0])\n",
    "blast_results['target'] = blast_results['target'].apply(lambda x: x.split('(')[0])\n",
    "blast_results = blast_results.sort_values(by=['query', 'eval'], ascending=[True, True])\n",
    "blast_results  = blast_results.drop_duplicates(subset='query', keep='first')\n",
    "\n",
    "def parse_fasta(file_path):\n",
    "    with open(file_path, \"r\") as f:\n",
    "        fasta_data = f.readlines()\n",
    "\n",
    "    # Initialize an empty list to store the results\n",
    "    results = {}\n",
    "\n",
    "    # Regular expressions for matching the labels\n",
    "    vfg_pattern = r\"VFG\\d{6}\"  # Match VFG followed by 6 digits\n",
    "    vf_pattern = r\"VF\\d{4,6}\"  # Match VF followed by 4 to 6 digits\n",
    "    vfc_pattern = r\"VFC\\d{4,6}\"  # Match VFC followed by 4 to 6 digits\n",
    "\n",
    "    # Iterate over the lines\n",
    "    for line in fasta_data:\n",
    "        # If the line starts with '>' it's a header\n",
    "        if line.startswith(\">\"):\n",
    "            # Extract the VFG label using regex\n",
    "            vfg_match = re.search(vfg_pattern, line)\n",
    "            vf_match = re.search(vf_pattern, line)\n",
    "            vfc_match = re.search(vfc_pattern, line)\n",
    "\n",
    "            if vfg_match:\n",
    "                vfg_label = vfg_match.group(0)\n",
    "                \n",
    "                # Extract VF and VFC labels if they exist in the header\n",
    "                vf_label = vf_match.group(0) if vf_match else \"Unknown VF\"\n",
    "                vfc_label = vfc_match.group(0) if vfc_match else \"Unknown VFC\"\n",
    "                \n",
    "                results[vfg_label] =  (vf_label, vfc_label)\n",
    "\n",
    "    return results\n",
    "\n",
    "# Step 2: Use the function and print the results\n",
    "file_path = \"../../data/blastdbs/vfdb/VFDB_setA_pro.fas\"  # Replace with your FASTA file path\n",
    "mapped_data = parse_fasta(file_path)\n",
    "\n",
    "vf_annots = pd.read_csv('../../data/blastdbs/vfdb/VFs.csv', skiprows=1, index_col=0)\n",
    "\n",
    "blast_results['vfid'] = blast_results.target.apply(lambda x: mapped_data[x][0])\n",
    "blast_results['category'] = blast_results['vfid'].apply(lambda x: vf_annots.loc[x, 'VFcategory'])\n",
    "blast_results['vfname'] = blast_results['vfid'].apply(lambda x: vf_annots.loc[x, 'VF_Name'])\n",
    "blast_results['gene_name'] = [bakta_annotations.loc[x, 'Name'] for x in blast_results['query']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe562f9-9d59-4208-9025-c607011e0452",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_amr_genes = [x for x in amr['Protein identifier'] if x in df_rare.index]\n",
    "rare_vf_genes = [x for x in blast_results['query'] if x in df_rare.index]\n",
    "total_virulence_genes = (rare_amr_genes + rare_vf_genes)\n",
    "vir_gene_stats = pd.DataFrame(index = total_virulence_genes, columns=['dir_neighbors', 'num_strains', 'neighbors_list'])\n",
    "for gene in set(rare_amr_genes + rare_vf_genes):\n",
    "    direct_neighbors = [x for x in rare_genome_stats.loc[gene, 'all_neighbors'] if x in total_virulence_genes] # direct neighbors in any contig\n",
    "\n",
    "    num_strains = df_rare.loc[gene].sum()\n",
    "    vir_gene_stats.loc[gene] = [len(direct_neighbors),  num_strains, direct_neighbors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac6af70-4784-438e-81ec-4e452e0d15da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'df' is your DataFrame containing the 'num_neighbors' column\n",
    "df = vir_gene_stats\n",
    "\n",
    "# Create the histogram data\n",
    "hist_data, bins = np.histogram(df.num_neighbors, bins=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])\n",
    "\n",
    "# Create a horizontal bar plot\n",
    "plt.figure(figsize=(6, 8))\n",
    "plt.barh(range(1, 12), hist_data, edgecolor='black')\n",
    "\n",
    "# Title and axis labels\n",
    "plt.title(\"Histogram of AMR/VF Rare Gene Neighborhoods\")\n",
    "plt.ylabel(\"Number of Direct Neighbors\")\n",
    "plt.xlabel(\"Frequencies\")\n",
    "\n",
    "plt.savefig('virulence_neighbor_barplot.svg', format='svg')\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0664915d-8ac5-4df9-abe4-20f8958fb005",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = vir_gene_stats\n",
    "bins = [0, 2, 10, float('inf')]  # Adjust these bins as necessary\n",
    "labels = [f\"Low ({bins[0]+1}-{bins[1]} strains)\", f\"Medium ({bins[1]+1}-{bins[2]} strains)\", f\"High ({bins[2]+1}+ strains)\"]\n",
    "df['Strain Category'] = pd.cut(df['num_strains'], bins=bins, labels=labels)\n",
    "\n",
    "# Create the box plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.boxplot(x=\"Strain Category\", y=\"dir_neighbors\", data=df, palette=\"pastel\")\n",
    "plt.title(\"Distribution of Direct Neighbors Across Strain Categories\")\n",
    "plt.xlabel(\"Strain Category\")\n",
    "plt.ylabel(\"Number of Direct Neighbors\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f52645e-c889-4e24-b5bb-5c95f4a2ff35",
   "metadata": {},
   "outputs": [],
   "source": [
    "vir_gene_stats['num_neighbors'] = vir_gene_stats.neighbors_list.apply(lambda x: len(x))\n",
    "vir_gene_stats.sort_values(by=['num_strains', 'num_neighbors'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d130eb43-21cf-41c1-be7e-3fcc213aeff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "\n",
    "# Example data (replace with your own dataframe)\n",
    "df = vir_gene_stats  # Assuming 'vir_gene_stats' is your dataframe\n",
    "\n",
    "# Count the occurrences of each (dir_neighbors, num_strains) pair\n",
    "count_df = df.groupby(['dir_neighbors', 'num_strains']).size().reset_index(name='count')\n",
    "count_df['count'] = np.log2(count_df['count'] + 1)\n",
    "\n",
    "# Scatter plot with color based on the count of each pair\n",
    "plt.figure(figsize=(8, 8))\n",
    "\n",
    "# Create scatter plot and store the result in the variable\n",
    "scatter = sns.scatterplot(x='dir_neighbors', y='num_strains', sizes=(50, 200), \n",
    "                          hue='count', data=count_df, palette='viridis', legend=False)\n",
    "\n",
    "# Add a colorbar based on the 'count' hue\n",
    "norm = mpl.colors.Normalize(vmin=count_df['count'].min(), vmax=count_df['count'].max())  # Normalize the color range\n",
    "sm = plt.cm.ScalarMappable(cmap='viridis', norm=norm)  # Create the mappable object for colorbar\n",
    "sm.set_array([])  # Empty array because the data is already embedded in the plot\n",
    "\n",
    "# Pass the current axis (ax) to colorbar\n",
    "plt.colorbar(sm, ax=scatter, label='Log2 Frequency')\n",
    "\n",
    "# Plot titles and labels\n",
    "plt.title('Scatter Plot of dir_neighbors vs num_strains with Frequency-Based Coloring')\n",
    "plt.xlabel('dir_neighbors')\n",
    "plt.ylabel('num_strains')\n",
    "\n",
    "plt.savefig('virulence_neighbor_scatter.svg', format='svg')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09bdb7c1-085b-4617-aa32-4e431892cb0f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-10T07:42:57.613564Z",
     "iopub.status.busy": "2024-12-10T07:42:57.613374Z",
     "iopub.status.idle": "2024-12-10T07:42:57.616146Z",
     "shell.execute_reply": "2024-12-10T07:42:57.615660Z",
     "shell.execute_reply.started": "2024-12-10T07:42:57.613553Z"
    }
   },
   "source": [
    "# Find locus of genes of interest for yersiniabactin production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3003197b-9164-4916-b492-a6ab83d0e0e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genome_and_locus = set([x for xs in rare_genome_stats.loc[vir_gene_stats.sort_values('dir_neighbors').tail(11).index].genomes.values for x in xs])\n",
    "for genome, locus in [x.split('_') for x in genome_and_locus]:\n",
    "    strain_gff = strain_vectors[genome]\n",
    "    contig_locus = strain_contigs[genome][int(locus)]\n",
    "    \n",
    "    start = strain_gff[strain_gff.gene.isin(contig_locus)].iloc[0,1]\n",
    "    end = strain_gff[strain_gff.gene.isin(contig_locus)].iloc[-1,2] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e283ed5-db0e-4e91-b435-daac327fd91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rgba_to_hex(rgba):\n",
    "    # Extract RGBA components\n",
    "    r, g, b, a = rgba\n",
    "    \n",
    "    # Convert RGB to range [0, 255]\n",
    "    r = int(r * 255)\n",
    "    g = int(g * 255)\n",
    "    b = int(b * 255)\n",
    "    \n",
    "    # Convert to hex string and return\n",
    "    return f\"#{r:02x}{g:02x}{b:02x}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d20ef2f-e24b-4d11-bc74-8e63584ab5ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_and_locus = [\n",
    "    '158836.2032_100',\n",
    "    '158836.1457_54',\n",
    "    '2850492.3_41',\n",
    "    '158836.1536_54',\n",
    "    '158836.1537_54',\n",
    "    '550.2451_23',\n",
    "    '158836.434_43',\n",
    "    '158836.416_44',\n",
    "    '2077137.3_218'\n",
    "] # same strains in genome and locus set but re-ordered, not necessary for plot but improves visual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2cc4705-6c9d-4ca0-8097-281c3734a293",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "\n",
    "\n",
    "name = \"test_fig\"\n",
    "gd_diagram = GenomeDiagram.Diagram(name)\n",
    "# set max len\n",
    "max_len = 0\n",
    "gene_counts = Counter()\n",
    "for genome, locus in [x.split('_') for x in genome_and_locus]:\n",
    "    strain_gff = strain_vectors[genome]\n",
    "    contig_locus = strain_contigs[genome][int(locus)]\n",
    "    strain_gff = strain_gff[strain_gff.gene.isin(contig_locus)]\n",
    "\n",
    "    \n",
    "    for gene in strain_gff.gene.unique():\n",
    "        gene_counts[gene] += 1\n",
    "    \n",
    "    max_len = max(max_len, strain_gff.end.max() - strain_gff.start.min())\n",
    "\n",
    "# Normalize the gene counts to a range of 0-1\n",
    "max_count = max(gene_counts.values()) if gene_counts else 1  # Prevent division by zero\n",
    "normalized_counts = {gene: count / max_count for gene, count in gene_counts.items()}\n",
    "\n",
    "\n",
    "# Create a custom colormap from white to orange\n",
    "cmap = LinearSegmentedColormap.from_list(\"gradient\", [(.7, .7, .7), (0, 0, 0)])\n",
    "\n",
    "# Function to get the color based on the normalized count\n",
    "def get_color_from_count(count):\n",
    "    norm_count = normalized_counts.get(count, 0)  # Default to 0 if the gene is not in the counts\n",
    "    return rgba_to_hex(cmap(norm_count))\n",
    "\n",
    "for genome, locus in [x.split('_') for x in genome_and_locus]:\n",
    "    \n",
    "    strain_gff = strain_vectors[genome]\n",
    "    contig_locus = strain_contigs[genome][int(locus)]\n",
    "    strain_gff = strain_gff[strain_gff.gene.isin(contig_locus)]\n",
    "     # Set colors: red for virulence factor, otherwise a gradient based on count\n",
    "    colors = []\n",
    "    for gene in strain_gff.gene.values:\n",
    "        if gene in vir_gene_stats.index:\n",
    "            colors.append('#FF0000')  # Red for virulence factors\n",
    "        else:\n",
    "            color = get_color_from_count(gene)  # Get the gradient color for the gene\n",
    "            colors.append(color)\n",
    "\n",
    "    \n",
    "    genome_region_len = strain_gff.end.max() - strain_gff.start.min()\n",
    "    offset = int((max_len - genome_region_len) / 2)\n",
    "    \n",
    "    if strain_gff[strain_gff.gene == 'Ebacter_C102'].strand.iloc[0] != '+':\n",
    "        strain_gff = strain_gff.iloc[::-1]\n",
    "        colors = colors[::-1]\n",
    "\n",
    "        min_coord = strain_gff[\"start\"].min()\n",
    "        max_coord = strain_gff[\"end\"].max()\n",
    "\n",
    "        strain_gff_temp = strain_gff.copy()\n",
    "        strain_gff_temp[\"start\"] = max_coord - (strain_gff[\"end\"] - min_coord)\n",
    "        strain_gff_temp[\"end\"] = max_coord - (strain_gff[\"start\"] - min_coord)\n",
    "        strain_gff_temp['strand'] = ['-' if x == '+' else '+' for x in strain_gff_temp.strand]\n",
    "        strain_gff = strain_gff_temp\n",
    "\n",
    "    track_name = str(genome) + '\\t' + str(metadata.set_index('genome_id').loc[genome, 'isolation_country']) + '\\t' + \\\n",
    "        str(metadata.set_index('genome_id').loc[genome, 'collection_date'])[:4] + '\\n' + A_binarized.loc[characterized_order, genome].idxmax()\n",
    "    gd_track_for_features = gd_diagram.new_track(\n",
    "            1, name=track_name, greytrack=True, start=0, end=max_len, height = .5, greytrack_labels = 1\n",
    "    )\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "    initial_locus = strain_gff.start.min()\n",
    "    # Add features to the track\n",
    "    for i, (index, row) in enumerate(strain_gff.iterrows()):\n",
    "        start = int(row['start'] - initial_locus) + offset\n",
    "        end = int(row['end'] - initial_locus) + offset\n",
    "        gene_name = row['gene']\n",
    "        color = colors[i]\n",
    "        strand = +1\n",
    "        if row.strand == '-':\n",
    "            strand = -1\n",
    "        \n",
    "        feature = SeqFeature(\n",
    "            FeatureLocation(start, end),\n",
    "            type=\"gene\",\n",
    "            qualifiers={\"gene\": gene_name}\n",
    "        )\n",
    "        feature.strand = strand\n",
    "        \n",
    "        gd_feature_set.add_feature(\n",
    "            feature,\n",
    "            sigil=\"BIGARROW\",\n",
    "            color=color,\n",
    "            label=False,\n",
    "            name=gene_name,\n",
    "            label_position=\"start\",\n",
    "            label_size=6,\n",
    "            label_angle=0,\n",
    "        )\n",
    "\n",
    "\n",
    "from Bio.Graphics.GenomeDiagram import CrossLink\n",
    "from reportlab.lib import colors\n",
    "\n",
    "\n",
    "for i in range(len(gd_diagram.tracks)-1):\n",
    "    tracks1 = gd_diagram.tracks[i+1]\n",
    "    tracks2 = gd_diagram.tracks[i+2]\n",
    "\n",
    "    for feature1 in tracks1.get_sets()[0].get_features():\n",
    "        for feature2 in tracks2.get_sets()[0].get_features():\n",
    "            if feature1.name == feature2.name:\n",
    "                color = colors.Color(red=(240/255),green=(240/255),blue=(240/255))\n",
    "                color2 = colors.Color(red=(240/255),green=(240/255),blue=(240/255))\n",
    "                if feature1.name in vir_gene_stats.index:\n",
    "                    color = colors.Color(red=(255/255),green=(200/255),blue=(200/255))\n",
    "                    color2  = colors.white\n",
    "                link_xy = CrossLink(\n",
    "                    (tracks1, feature1.location.start, feature1.location.end),\n",
    "                    (tracks2, feature2.location.start, feature2.location.end),\n",
    "                    color,\n",
    "                    color2,\n",
    "                )\n",
    "                gd_diagram.cross_track_links.append(link_xy)\n",
    "                \n",
    "\n",
    "gd_diagram.draw(format=\"linear\", pagesize=\"A4\", fragments=1, start=0, end=max_len)\n",
    "gd_diagram.write(name + \".svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7136542d-1526-46b7-a265-731ce318024d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python phylon_analysis",
   "language": "python",
   "name": "phylon_analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
